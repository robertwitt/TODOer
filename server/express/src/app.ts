import express from "express";
import YAML from "yamljs";
import path from "path";
import { connector } from "swagger-routes-express";
import * as apiHandlers from "./api";
import { handleError } from "./middleware/error";

const app = express();

app.set("port", process.env.PORT || 3000);

// Setup body parsers
app.use(express.json());
app.use(express.text());
app.use(express.urlencoded({ extended: false }));

// Connect API and app
const apiDefinition = YAML.load(path.join("resources", "api.yaml"));
const connect = connector(apiHandlers, apiDefinition);
connect(app);

// Add error handler
app.use(handleError);

export default app;
