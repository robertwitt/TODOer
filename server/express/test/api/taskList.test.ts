import supertest from "supertest";
import app from "../../src/app";
import MockDb from "../../src/repository/mock/db";

describe("The API server", () => {
  const request = supertest(app);

  beforeEach(() => MockDb.instance.reset());

  it("can read a single task list by ID", async () => {
    const { body } = await request.get("/beta/TaskLists/1").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("cannot read a task list by an unknown ID", async () => {
    await request.get("/beta/TaskLists/99").expect(404);
  });

  it("can find all task lists", async () => {
    const { body } = await request.get("/beta/TaskLists").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can read a single task list with all its tasks", async () => {
    const { body } = await request
      .get("/beta/TaskLists/1/assignedTasks")
      .expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can read the My Day view with all its tasks", async () => {
    const { body } = await request
      .get("/beta/TaskLists/2/assignedTasks")
      .expect(200);
    expect(body).toMatchSnapshot();
  });

  it("can create a new task list", async () => {
    const { body } = await request
      .post("/beta/TaskLists")
      .send({
        title: "new list",
        color: "92FA4E",
      })
      .expect(201);
    expect(body).toMatchSnapshot();
  });

  it("can create a new task list as default", async () => {
    await request
      .post("/beta/TaskLists")
      .send({
        title: "new list",
        isDefaultCollection: true,
      })
      .expect(201);
    const { body } = await request.get("/beta/TaskLists").expect(200);
    expect(body).toMatchSnapshot();
  });

  it("cannot create a task list with invalid", async () => {
    await request.post("/beta/TaskLists").send({ color: "WDR392" }).expect(400);
  });
});
