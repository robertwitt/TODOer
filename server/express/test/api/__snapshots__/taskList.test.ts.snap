// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The API server can create a new task list 1`] = `
Object {
  "color": "92FA4E",
  "id": 43,
  "isDefaultCollection": false,
  "isDeletable": true,
  "isUpdatable": true,
  "title": "new list",
  "type": "COLLECTION",
}
`;

exports[`The API server can create a new task list as default 1`] = `
Object {
  "value": Array [
    Object {
      "color": null,
      "id": 2,
      "isDefaultCollection": false,
      "isDeletable": false,
      "isUpdatable": false,
      "title": "My Day",
      "type": "MY_DAY",
    },
    Object {
      "color": null,
      "id": 3,
      "isDefaultCollection": false,
      "isDeletable": false,
      "isUpdatable": false,
      "title": "Tomorrow",
      "type": "TOMORROW",
    },
    Object {
      "color": "0000FF",
      "id": 1,
      "isDefaultCollection": false,
      "isDeletable": true,
      "isUpdatable": true,
      "title": "My Tasks",
      "type": "COLLECTION",
    },
    Object {
      "color": null,
      "id": 43,
      "isDefaultCollection": true,
      "isDeletable": true,
      "isUpdatable": true,
      "title": "new list",
      "type": "COLLECTION",
    },
    Object {
      "color": null,
      "id": 42,
      "isDefaultCollection": false,
      "isDeletable": true,
      "isUpdatable": true,
      "title": null,
      "type": "COLLECTION",
    },
  ],
}
`;

exports[`The API server can find all task lists 1`] = `
Object {
  "value": Array [
    Object {
      "color": null,
      "id": 2,
      "isDefaultCollection": false,
      "isDeletable": false,
      "isUpdatable": false,
      "title": "My Day",
      "type": "MY_DAY",
    },
    Object {
      "color": null,
      "id": 3,
      "isDefaultCollection": false,
      "isDeletable": false,
      "isUpdatable": false,
      "title": "Tomorrow",
      "type": "TOMORROW",
    },
    Object {
      "color": "0000FF",
      "id": 1,
      "isDefaultCollection": true,
      "isDeletable": true,
      "isUpdatable": true,
      "title": "My Tasks",
      "type": "COLLECTION",
    },
    Object {
      "color": null,
      "id": 42,
      "isDefaultCollection": false,
      "isDeletable": true,
      "isUpdatable": true,
      "title": null,
      "type": "COLLECTION",
    },
  ],
}
`;

exports[`The API server can read a single task list by ID 1`] = `
Object {
  "color": "0000FF",
  "id": 1,
  "isDefaultCollection": true,
  "isDeletable": true,
  "isUpdatable": true,
  "title": "My Tasks",
  "type": "COLLECTION",
}
`;

exports[`The API server can read a single task list with all its tasks 1`] = `
Object {
  "assignedTasks": Array [
    Object {
      "collection": Object {
        "id": 1,
        "title": "My Tasks",
      },
      "dueDate": "2021-05-31",
      "dueTime": null,
      "id": 1,
      "isDeletable": true,
      "isPlannedForMyDay": false,
      "isUpdatable": true,
      "priority": Object {
        "code": 5,
        "name": "low",
      },
      "status": Object {
        "code": "O",
        "name": "open",
      },
      "title": "Tax declaration",
    },
    Object {
      "collection": Object {
        "id": 1,
        "title": "My Tasks",
      },
      "dueDate": null,
      "dueTime": null,
      "id": 2,
      "isDeletable": true,
      "isPlannedForMyDay": true,
      "isUpdatable": true,
      "priority": Object {
        "code": 5,
        "name": "low",
      },
      "status": Object {
        "code": "O",
        "name": "open",
      },
      "title": "Groceries shopping",
    },
  ],
  "color": "0000FF",
  "id": 1,
  "title": "My Tasks",
  "type": "COLLECTION",
}
`;

exports[`The API server can read the My Day view with all its tasks 1`] = `
Object {
  "assignedTasks": Array [
    Object {
      "collection": Object {
        "id": 1,
        "title": "My Tasks",
      },
      "dueDate": null,
      "dueTime": null,
      "id": 2,
      "isDeletable": true,
      "isPlannedForMyDay": true,
      "isUpdatable": true,
      "priority": Object {
        "code": 5,
        "name": "low",
      },
      "status": Object {
        "code": "O",
        "name": "open",
      },
      "title": "Groceries shopping",
    },
  ],
  "color": null,
  "id": 2,
  "title": "My Day",
  "type": "MY_DAY",
}
`;
